core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -941,39 +857,129 @@ class ReplicaManager(val config: KafkaConfig,
+      // We verify above that all partitions use the same producer ID.
+      val batchInfo = transactionVerificationEntries.unverified.head._2.firstBatch()
+      addPartitionsToTxnManager.foreach(_.verifyTransaction(
+        transactionalId = transactionalId,
+        producerId = batchInfo.producerId,
+        producerEpoch = batchInfo.producerEpoch,
+        topicPartitions = transactionVerificationEntries.unverified.keySet.toSeq,
+        callback = KafkaRequestHandler.wrapAsyncCallback(onVerificationComplete, requestLocal)
+      ))
+    }
+  }
+
+  /**
+   * A helper method to compile the results from the transaction verification and call the postVerificationCallback.
+   *
+   * @param transactionVerificationEntries the object that will store the entries to verify, the errors, and the verification guards
+   * @param postVerificationCallback       the method to be called when verification completes and the verification errors
+   *                                       and the thread's RequestLocal are supplied
+   * @param requestLocal                   container for the stateful instances scoped to this request -- this must correspond to the
+   *                                       thread calling this method
+   *
+   */
+  private def executePostVerificationCallback(transactionVerificationEntries: TransactionVerificationEntries,
+                                                      postVerificationCallback: RequestLocal => Map[TopicPartition, LogAppendResult] => Unit)
+                                                     (requestLocal: RequestLocal, unverifiedEntries: Map[TopicPartition, Errors]): Unit = {
+    val errorResults = (unverifiedEntries ++ transactionVerificationEntries.errors).map {
+      case (topicPartition, error) =>
+        // translate transaction coordinator errors to known producer response errors
+        val customException =
+          error match {
+            case Errors.INVALID_TXN_STATE => Some(error.exception("Partition was not added to the transaction"))
+            case Errors.CONCURRENT_TRANSACTIONS |
+                 Errors.COORDINATOR_LOAD_IN_PROGRESS |
+                 Errors.COORDINATOR_NOT_AVAILABLE |
+                 Errors.NOT_COORDINATOR => Some(new NotEnoughReplicasException(
+              s"Unable to verify the partition has been added to the transaction. Underlying error: ${error.toString}"))
+            case _ => None
+          }
+        topicPartition -> LogAppendResult(
+          LogAppendInfo.UNKNOWN_LOG_APPEND_INFO,
+          Some(customException.getOrElse(error.exception)),
+          hasCustomErrorMessage = customException.isDefined
+        )
+    }
+    postVerificationCallback(requestLocal)(errorResults)
+  }
+
+  private def partitionEntriesForVerification(transactionVerificationEntries: TransactionVerificationEntries, entriesPerPartition: Map[TopicPartition, MemoryRecords]): TransactionVerificationEntries = {
```

Comment:
Not now. I played around with a few ways of doing this -- one was to create the object here. However, I needed the guards for the post verification callback. I can make this return Unit to avoid confusion.

Reply:
The new code also removes this.